From: Erik de Castro Lopo <erikd@mega-nerd.com>
Date: Thu, 11 Dec 2014 09:40:37 +0000 (+1100)
Subject: stream_decoder.c : Fix free-ing of unitilialized pointer.
X-Git-Url: https://git.xiph.org/?p=flac.git;a=commitdiff_plain;h=18a485451014b766573fc104fc8bfc57c61d822f

stream_decoder.c : Fix free-ing of unitilialized pointer.

Found using afl (http://lcamtuf.coredump.cx/afl/).

index d599580..d13b23b 100644
--- src/libFLAC/stream_decoder.c
+++ src/libFLAC/stream_decoder.c
@@ -1737,6 +1737,10 @@
 				return false;
 			}
 			for (i = 0; i < obj->num_comments; i++) {
+				/* Initialize here just to make sure. */
+				obj->comments[i].length = 0;
+				obj->comments[i].entry = 0;
+
 				FLAC__ASSERT(FLAC__STREAM_METADATA_VORBIS_COMMENT_ENTRY_LENGTH_LEN == 32);
 				if (length < 4) {
 					obj->num_comments = i;
@@ -1748,8 +1752,6 @@
 					return false; /* read_callback_ sets the state for us */
 				if (obj->comments[i].length > 0) {
 					if (length < obj->comments[i].length) {
-						obj->comments[i].length = 0;
-						obj->comments[i].entry = 0;
 						obj->num_comments = i;
 						goto skip;
 					}
@@ -1759,8 +1761,11 @@
 						decoder->protected_->state = FLAC__STREAM_DECODER_MEMORY_ALLOCATION_ERROR;
 						return false;
 					}
-					if (!FLAC__bitreader_read_byte_block_aligned_no_crc(decoder->private_->input, obj->comments[i].entry, obj->comments[i].length))
-						return false; /* read_callback_ sets the state for us */
+					memset (obj->comments[i].entry, 0, obj->comments[i].length) ;
+					if (!FLAC__bitreader_read_byte_block_aligned_no_crc(decoder->private_->input, obj->comments[i].entry, obj->comments[i].length)) {
+						obj->num_comments = i;
+						goto skip;
+					}
 					obj->comments[i].entry[obj->comments[i].length] = '\0';
 				}
 				else
